### PR TITLE
[EKS] Update autoscaler image

### PR DESCRIPTION
**Pull request by amandineslx**
[sc-145288]

### CODE REVIEW

**Review by marchelleboid — CHANGES_REQUESTED** [review]
Tested with multiple k8s versions and the default version, and a correct version of the autoscaler was added each time. It's too bad that we need a version mapping that we'll need to continually maintain, but there doesn't seem to be a better "automatic" option.

File: python-lib/dku_kube/autoscaler.py [hunk 1]
```diff
@@ -1,26 +1,46 @@
...
     env = os.environ.copy()
     env['KUBECONFIG'] = kube_config_path
     cmd = ['kubectl', 'get', 'pods', '--namespace', 'kube-system', '-l', 'app=cluster-autoscaler', '--ignore-not-found']
-    logging.info("Create autoscaler with : %s" % json.dumps(cmd))
+    logging.info("Checking autoscaler presence with : %s" % json.dumps(cmd))
     out, err = run_with_timeout(cmd, env=env, timeout=5)
     return len(out.strip()) > 0
 
-def add_autoscaler_if_needed(cluster_id, kube_config_path):
+def add_autoscaler_if_needed(cluster_id, cluster_config, kube_config_path, kubernetes_version):
```
[the remark this answers is not in the document]
Comment (amandineslx) [reply python-lib/dku_kube/autoscaler.py:26]: done

File: python-lib/dku_kube/autoscaler.py [hunk 2]
```diff
@@ -1,26 +1,46 @@
...
+import os, json, logging
 from .kubectl_command import run_with_timeout
+from dku_utils.access import _is_none_or_blank
+from dku_utils.tools_version import get_kubernetes_default_version
+
+AUTOSCALER_IMAGES = {
+    '1.20': 'k8s.gcr.io/autoscaling/cluster-autoscaler:v1.20.3',
+    '1.21': 'k8s.gcr.io/autoscaling/cluster-autoscaler:v1.21.2',
+    '1.22': 'k8s.gcr.io/autoscaling/cluster-autoscaler:v1.22.2',
+    '1.23': 'k8s.gcr.io/autoscaling/cluster-autoscaler:v1.23.1',
```
[the remark this answers is not in the document]
Comment (amandineslx) [reply python-lib/dku_kube/autoscaler.py:10]: cut-off put at 1.24 which resolves the different image sources

File: python-lib/dku_kube/autoscaler.py [hunk 3]
```diff
@@ -1,26 +1,46 @@
...
-def add_autoscaler_if_needed(cluster_id, kube_config_path):
+def add_autoscaler_if_needed(cluster_id, cluster_config, kube_config_path, kubernetes_version):
     if not has_autoscaler(kube_config_path):
+        if _is_none_or_blank(kubernetes_version):
+            kubernetes_version = get_kubernetes_default_version(cluster_config)
         autoscaler_file_path = 'autoscaler.yaml'
+        if float(kubernetes_version) < 1.20:
+          autoscaler_image = AUTOSCALER_IMAGES.get('1.20', 'k8s.gcr.io/autoscaling/cluster-autoscaler:v1.20.3')
+        else:  
+          autoscaler_image = AUTOSCALER_IMAGES.get(kubernetes_version, 'registry.k8s.io/autoscaling/cluster-autoscaler:v1.28.0')
```
[the remark this answers is not in the document]
Comment (amandineslx) [reply python-lib/dku_kube/autoscaler.py:34]: done

File: python-lib/dku_kube/autoscaler.py [hunk 4]
```diff
@@ -1,26 +1,46 @@
...
     out, err = run_with_timeout(cmd, env=env, timeout=5)
     return len(out.strip()) > 0
 
-def add_autoscaler_if_needed(cluster_id, kube_config_path):
+def add_autoscaler_if_needed(cluster_id, cluster_config, kube_config_path, kubernetes_version):
     if not has_autoscaler(kube_config_path):
+        if _is_none_or_blank(kubernetes_version):
+            kubernetes_version = get_kubernetes_default_version(cluster_config)
         autoscaler_file_path = 'autoscaler.yaml'
+        if float(kubernetes_version) < 1.20:
```
[the remark this answers is not in the document]
Comment (amandineslx) [reply python-lib/dku_kube/autoscaler.py:31]: done

[the remark this answers is not in the document]
Comment (amandineslx) [reply python-lib/dku_kube/autoscaler.py:31]: it is not necessary as the values that can be provided in the `Kubernetes version` field in the cluster creation form are restricted

Screenshot 2023-11-17 at 15 52 30

File: python-lib/dku_utils/tools_version.py [hunk 5]
```diff
@@ -40,3 +42,8 @@ def check_versions():
     if kubectl_should_use_beta_apiVersion(kubectl_version) and authenticator_version < '0.5.4':
         raise Exception('Found kubectl %s and aws-iam-authenticator %s, which are incompatible. Please upgrade aws-iam-authenticator.' 
                         % (kubectl_version['major']+'.'+(kubectl_version['minor']), authenticator_version))
+
+def get_kubernetes_default_version(cluster_config):
+    cmd = EksctlCommand(['utils', 'schema'], get_connection_info(cluster_config))
+    out = cmd.run_and_get_output()
+    return json.loads(out)['definitions']['ClusterMeta']['properties']['version']['default']
```
[the remark this answers is not in the document]
Comment (amandineslx) [reply python-lib/dku_utils/tools_version.py]: first fix attempt pushed but there is still an issue with using the K8S default version and I am investigating

File: python-lib/dku_kube/autoscaler.py [hunk 6]
```diff
@@ -1,26 +1,46 @@
...
     cmd = ['kubectl', 'get', 'pods', '--namespace', 'kube-system', '-l', 'app=cluster-autoscaler', '--ignore-not-found']
-    logging.info("Create autoscaler with : %s" % json.dumps(cmd))
+    logging.info("Checking autoscaler presence with : %s" % json.dumps(cmd))
     out, err = run_with_timeout(cmd, env=env, timeout=5)
     return len(out.strip()) > 0
 
-def add_autoscaler_if_needed(cluster_id, kube_config_path):
+def add_autoscaler_if_needed(cluster_id, cluster_config, cluster_def, kube_config_path):
     if not has_autoscaler(kube_config_path):
+        print("Amandine K8S version: %s" % cluster_config.get("k8sVersion", None))
```
Comment (marchelleboid) [inline python-lib/dku_kube/autoscaler.py:24]: Can be removed, I think (and the one a couple lines down).

File: python-lib/dku_utils/tools_version.py [hunk 7]
```diff
@@ -23,6 +32,18 @@ def get_kubectl_version_int(kubectl_version):
...
+def strip_kubernetes_version(k8s_version_input):
+    """
+    Removes any additional characters from the Kubernetes version specified in the cluster creation form
+    """
+    regex_k8s_version = re.compile("^[^0-9]*([0-9]+\.?[0-9]+)([^0-9].*$|$)")
+    search_results_k8s_version = re.search(regex_k8s_version, k8s_version_input)
+    if not search_results_k8s_version or not search_results_k8s_version.groups():
+        raise Exception("Kubectl version specified: %s. No valid Kubernetes version found", k8s_version_input)
+    for g in search_results_k8s_version.groups():
+        print("Amandine capture group: %s" % g)
```
Comment (marchelleboid) [inline python-lib/dku_utils/tools_version.py:44]: Another one here

File: python-lib/dku_kube/autoscaler.py [hunk 8]
```diff
@@ -1,26 +1,46 @@
...
+        kubernetes_version = cluster_config.get("k8sVersion", None)
+        if _is_none_or_blank(kubernetes_version):
+            kubernetes_version = cluster_def.get("Version")
+        kubernetes_version = strip_kubernetes_version(kubernetes_version)
```
Comment (marchelleboid) [inline python-lib/dku_kube/autoscaler.py:28]: Now that we don't make an extra call, we could always just take the `cluster_def.get("Version")` version, but I think it's fine to leave it like this as well.

**Review by marchelleboid — APPROVED** [review]
Retested with leaving the version blank, and the correct k8s version was pulled from the json (and the correct autoscaler installed).

A couple debug statements that can be removed, but the code LGTM to me. Do we know yet how we will keep the autoscaler version mapping up to date?

File: python-lib/dku_kube/autoscaler.py (hunk 6 above)
Comment (amandineslx) [reply python-lib/dku_kube/autoscaler.py:24]: of course!

**Comment by amandineslx** [comment]
> Retested with leaving the version blank, and the correct k8s version was pulled from the json (and the correct autoscaler installed).
> 
> A couple debug statements that can be removed, but the code LGTM to me. Do we know yet how we will keep the autoscaler version mapping up to date?

We will discuss it internally

**Pull request merged**